### PR TITLE
Audigent HaloID User Id System Module

### PR DESCRIPTION
**Pull request by anthonylauzon**
## Type of change

- [X] Feature

## Description of change
This changeset adds support for the Audigent HaloID.

- contact email of the adapter’s maintainer
ant@audigent.com

## Other information
Audigent Segmentation Real-time Data Provider #4834
https://github.com/prebid/Prebid.js/pull/4834

### CODE REVIEW

**Comment by pm-harshad-mane** [comment]
Hello @anthonylauzon ,
Could you please add entries for HaloID in https://github.com/prebid/Prebid.js/blob/master/modules/userId/eids.md and https://github.com/prebid/Prebid.js/blob/master/modules/userId/eids.js This will help bidders like PrebidServerBid Adapter and PubMatic bid adapter to support HaloID without making any change in their code.

Also we will need to add entry in https://github.com/prebid/Prebid.js/blob/master/modules/.submodules.json to auto-select the parent module at build time.

File: modules/haloIdSystem.js
Comment (pm-harshad-mane) [inline modules/haloIdSystem.js:28]: In example storage type is mentioned as cookie but no cookie is set in the code.
Looks like HaloID is being set in local-storage even when storage type is cookie.
@idettman can pub control the local-storage usage through PBJS?

Comment (idettman) [reply modules/haloIdSystem.js:28]: Should use the methods in src/storageManager.js for local storage actions.

**Review by idettman — DISMISSED** [review]
Please use the storageManager.js functions for any local storage or cookie actions

**Comment by patmmccann** [comment]
there are no unit tests added to the user id module unit tests as all other prs have done

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:31]: external scripts/libraries aren't generally allowed: https://docs.prebid.org/dev-docs/module-rules#global-module-rules

File: modules/.submodules.json
Comment (smenzer) [inline modules/.submodules.json:23]: should this be in a separate PR since it seems unrelated to the user id module submission?

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:42]: loading external scripts are generally not allowed per the module rules: https://docs.prebid.org/dev-docs/module-rules#global-module-rules

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:57]: this isn't documented...are you expecting publishers to use this? or is it only for your test?

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:27]: what's the difference between these two local storage names, `haloId` and `auHaloId`?

**Review by smenzer — CHANGES_REQUESTED** [review]
A few comments about how you're retrieving and storing your ID from cache:

Your code is manually retrieving your ID _only_ from `localStorage` and _only_ from one of two names: `haloId` or `auHaloId`. However, are using the userId module to handle writing the ID to cache, which relies on the `storage` configuration settings, which you allow the publisher to configure (both the storage type and name). It's entirely possible that neither of the locations you hardcoded in your IdSystem contains the ID that was stored by the userId module.

I would suggest you either read the parameters from the publisher configuration to read the ID (or better yet, just add a third parameter to your `getId()` method which contains the cached ID, if any; or you need to write the ID to cache yourself so you know where it is located, and if you do, you should drop any `storage` config from the docs so publishers aren't confused.

File: modules/haloIdSystem.js
Comment (anthonylauzon) [reply modules/haloIdSystem.js:57]: it's just for a unit test

**Comment by patmmccann** [comment]
Appears to be missing changes to eids spec

File: modules/userId/eids.js
Comment (patmmccann) [inline modules/userId/eids.js:115]: if you guys look at the liveintent object, segments are coming on the eid. does it make sense for yours to mirror this structure?

File: modules/haloIdSystem.js
Comment (smenzer) [reply modules/haloIdSystem.js:57]: ok. i think there are probably easier ways to get an ID back for testing purposes (you can check out some of the other modules), but if it's only internal then it's more of a coding style than anything. i'll let one of the other reviewers weigh in on if there's a need to change this or not. thanks.

**Comment by smenzer** [comment]
i don't see any docs PRs for this, which are required before merging

**Comment by patmmccann** [comment]
@anthonylauzon Looks like another merge created conflicts, fyi

File: modules/userId/eids.js
Comment (anthonylauzon) [reply modules/userId/eids.js:115]: we intend to add segment functionality to our id endpoint with a further update.. this will allow for better integration with our rtd module and allow for fewer requests

File: modules/.submodules.json
Comment (anthonylauzon) [reply modules/.submodules.json:23]: the user id submodule and rtd probider are going to be linked in a future pr

**Comment by pm-harshad-mane** [comment]
Hello @anthonylauzon ,

Both checks, CI and LGTM are failing, can you please check?

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:2]: probably want to change this to your own name instead of The Trade Desk's

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:26]: name change

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:39]: this looks like a development url...is that really what you want?

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:59]: why are you getting all user ids and sending them to yourself?

File: modules/.submodules.json
Comment (smenzer) [inline modules/.submodules.json:24]: i still don't fully understand why this is part of the userid module, especially if there's no immediate need for the two modules to communicate.  should this at least be in a separate PR to make it clear?

**Review by smenzer — CHANGES_REQUESTED** [review]
some additional changes and a question

File: modules/haloIdSystem.js
Comment (anthonylauzon) [reply modules/haloIdSystem.js:59]: it will provide better id resolution

File: modules/.submodules.json
Comment (anthonylauzon) [reply modules/.submodules.json:24]: they will be communicating in a forthcoming PR

File: modules/haloIdSystem.js
Comment (patmmccann) [reply modules/haloIdSystem.js:59]: this violates rule 3 of id modules http://docs.prebid.org/dev-docs/module-rules.html

**Comment by patmmccann** [comment]
Gathering other ids violates rule 3 of id modules http://docs.prebid.org/dev-docs/module-rules.html

**Review by smenzer — CHANGES_REQUESTED** [review]
need to remove the call to `getUserIds()`

File: modules/haloIdSystem.js
Comment (anthonylauzon) [reply modules/haloIdSystem.js]: removed

Comment (smenzer) [reply modules/haloIdSystem.js]: thank you

File: modules/haloIdSystem.js
Comment (smenzer) [inline modules/haloIdSystem.js:37]: unnecessary comment since this isn't what the next line is doing

**Review by smenzer — COMMENTED** [review]
LGTM

**Review by smenzer — APPROVED** [review]
LGTM

File: modules/haloIdSystem.js
Comment (pm-harshad-mane) [reply modules/haloIdSystem.js:37]: @anthonylauzon can you please check the comment mentioned by @smenzer

File: test/spec/modules/userId_spec.js
Comment (pm-harshad-mane) [inline test/spec/modules/userId_spec.js:1228]: @anthonylauzon looks like we do not need this copied comment

**Comment by pm-harshad-mane** [comment]
@anthonylauzon have you also raised PR for the related documentation?
Refer https://github.com/prebid/prebid.github.io/blob/master/dev-docs/modules/userId.md

**Comment by anthonylauzon** [comment]
> @anthonylauzon have you also raised PR for the related documentation?
> Refer https://github.com/prebid/prebid.github.io/blob/master/dev-docs/modules/userId.md

**Comment by anthonylauzon** [comment]
> @anthonylauzon have you also raised PR for the related documentation?
> Refer https://github.com/prebid/prebid.github.io/blob/master/dev-docs/modules/userId.md
The documentation PR is located here:

https://github.com/prebid/prebid.github.io/pull/2320

**Comment by anthonylauzon** [comment]
@smenzer @pm-harshad-mane this is ready to go. can we please merge this so that we can begin integration with our RTD module?

**Comment by pm-harshad-mane** [comment]
Hello @jsnellbaker , @bretg 
Shall I merge this PR? 
Does everything looks OK to you guys?

**Review by pm-harshad-mane — APPROVED** [review]

**Review by smenzer — APPROVED** [review]
LGTM

**Pull request merged**
